File: Challenge/ViewController.m
//
//  ViewController.m
//  Challenge
//
//  Created by wanner on 18.09.13.
//  Copyright (c) 2013 Gerhard Wanner. All rights reserved.
//
//
// TODO
//

#import "ViewController.h"
#import <AudioToolbox/AudioToolbox.h>
#import "main.h"

SystemSoundID mClick;
SystemSoundID mBeepLow;
SystemSoundID mBeepHigh;

int mLap = -1;                      // Rundenanzahl
bool running = false;               // Gibt an, ob die Stopuhr läuft
bool inBox = false;                 // Gibt an, ob der Boxenstop bereits erfolgt ist
NSTimer *timer;                     // Timer für die asynchrone Zeitanzeige

NSDate *startTime;                  // Startzeit der Messung
NSDate *startLapTime;               // Startzeit der aktuellen Runde
NSDate *startRefTime;               // Startzeit der Referenzrunde
NSDate *endRefTime;                 // Endzeit der Referenzrunde

// Referenzrunde
NSTimeInterval refLapTimeInterval = 0;  // Zeit der Referenzrunde

const int ALERT_BOX = 0;
const int ALERT_RESET = 1;

@interface ViewController ()

@end

@implementation ViewController

@synthesize counter;
@synthesize totalTime;
@synthesize lapTime;
@synthesize lastLapTime;
@synthesize refLapTime;
@synthesize targetLapTime;
@synthesize targetTotalTime;
@synthesize boxButton;
@synthesize incButton;
@synthesize decButton;
@synthesize refPlusHButton;
@synthesize refMinusHButton;
@synthesize refPlusZButton;
@synthesize refMinusZButton;

-(void)alertView:(UIAlertView *) alertView clickedButtonAtIndex:(NSInteger)buttonIndex {
    switch (alertView.tag) {
        case (ALERT_BOX):
            if(buttonIndex != [alertView cancelButtonIndex]) {
                inBox = true;
                boxButton.enabled = false;
            }
            break;
        case ALERT_RESET:
            if(buttonIndex != [alertView cancelButtonIndex]) {
                [self reset];
            }
            break;
        default:
            break;
    }
}

-(IBAction)box:(id)sender
{
    AudioServicesPlaySystemSound(mClick);
    UIAlertView *alert = [[UIAlertView alloc]
                          initWithTitle:@"Box" message:@"Boxenstop?" delegate:self cancelButtonTitle:@"Nein" otherButtonTitles:@"Ja", nil ];
    alert.tag = ALERT_BOX;
    [alert show];
}

-(IBAction)dec:(id)sender
{
    mLap--;
    [self displayCounter];
    [self displayTargetLapTime];
    AudioServicesPlaySystemSound(mClick);
}

-(IBAction)refPlusH:(id)sender
{
    refLapTimeInterval = refLapTimeInterval+0.01;
    [self displayNewRefTime];
}

-(IBAction)refMinusH:(id)sender
{
    refLapTimeInterval = refLapTimeInterval-0.01;
    [self displayNewRefTime];
}

-(IBAction)refPlusZ:(id)sender
{
    refLapTimeInterval = refLapTimeInterval+0.1;
    [self displayNewRefTime];
}

-(IBAction)refMinusZ:(id)sender
{
    refLapTimeInterval = refLapTimeInterval-0.1;
    [self displayNewRefTime];
}

-(void)displayNewRefTime
{
    AudioServicesPlaySystemSound(mClick);
    [self displayRefTime];
    [self displayCounter];
    [self displayTargetLapTime];
    [self displayTargetTotalTime];
}

-(void)displayCounter
{
    int lap = mLap;
    if(lap == -1) {
        lap = 0;
    }
    counter.text = [NSString stringWithFormat:@"%d", lap ];
}

- (void)displayLapTime
{
    NSString * timeString;
    if(startTime != nil) {
        NSDate *currentDate = [NSDate date];
        NSTimeInterval timeInterval = [currentDate timeIntervalSinceDate:startLapTime];
        NSDate *timerDate = [NSDate dateWithTimeIntervalSince1970:timeInterval];
        NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
        [dateFormatter setDateFormat:@"m:ss.S"];
        [dateFormatter setTimeZone:[NSTimeZone timeZoneForSecondsFromGMT:0.0]];
        timeString = [dateFormatter stringFromDate:timerDate];
    } else {
        timeString = @"";
    }
    lapTime.text = timeString;
}

- (void)displayLastLapTime
{
    NSString * timeString;
    if(startLapTime != nil) {
        NSDate *currentDate = [NSDate date];
        NSTimeInterval timeInterval = [currentDate timeIntervalSinceDate:startLapTime];
        NSDate *timerDate = [NSDate dateWithTimeIntervalSince1970:timeInterval];
        NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
        [dateFormatter setDateFormat:@"m:ss.S"];
        [dateFormatter setTimeZone:[NSTimeZone timeZoneForSecondsFromGMT:0.0]];
        timeString = [dateFormatter stringFromDate:timerDate];
    } else {
        timeString = @"";
    }
    lastLapTime.text = timeString;
}

- (void)calculateRefTime
{
    if(startLapTime != nil) {
        refLapTimeInterval = [endRefTime timeIntervalSinceDate:startRefTime];
    }
}

- (void)displayRefTime
{
    NSString * timeString;
    if(refLapTimeInterval != 0) {
        NSDate *timerDate = [NSDate dateWithTimeIntervalSince1970:refLapTimeInterval];
        NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
        [dateFormatter setDateFormat:@"m:ss.SS"];
        [dateFormatter setTimeZone:[NSTimeZone timeZoneForSecondsFromGMT:0.0]];
        timeString = [dateFormatter stringFromDate:timerDate];
    } else {
        timeString = @"";
    }
    refLapTime.text = timeString;
    
}

-(void)displayTimerTimes
{
    [self displayTotalTime];
    [self displayTargetTotalTime];
    [self displayLapTime];
}

- (void)displayTotalTime
{
    NSString * timeString;
    if(startTime != nil) {
        NSDate *currentDate = [NSDate date];
        NSTimeInterval timeInterval = [currentDate timeIntervalSinceDate:startTime];
        NSDate *timerDate = [NSDate dateWithTimeIntervalSince1970:timeInterval];
        NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
        [dateFormatter setDateFormat:@"m:ss.S"];
        [dateFormatter setTimeZone:[NSTimeZone timeZoneForSecondsFromGMT:0.0]];
        timeString = [dateFormatter stringFromDate:timerDate];
    } else {
        timeString = @"0:00.0";
    }
    totalTime.text = timeString;
}

- (void)displayTargetLapTime
{
    NSString * timeString;
    
    if(mLap >= REF_LAP) {
        // Gesamtzeit Wertungsprüfung auf Basis der Referenzrunde
        NSTimeInterval timeInterval = (TOTAL_LAP-REF_LAP) * refLapTimeInterval;
        // Zeit seit Ende der Referenzrunde abziehen
        NSDate *currentDate = [NSDate date];
        timeInterval -= [currentDate timeIntervalSinceDate:endRefTime];
        // Vor Boxenstopp? --> Boxenzeit abziehen
        if(!inBox) {
            timeInterval -= BOX_TIME;
        }
        // Durch die Anzahl verbleibender Runden teilen
        timeInterval = timeInterval / (TOTAL_LAP-mLap);
        NSDate *timerDate = [NSDate dateWithTimeIntervalSince1970:timeInterval];
        NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
        [dateFormatter setDateFormat:@"m:ss.S"];
        [dateFormatter setTimeZone:[NSTimeZone timeZoneForSecondsFromGMT:0.0]];
        timeString = [dateFormatter stringFromDate:timerDate];
    } else {
        timeString = @"";
    }
    targetLapTime.text = timeString;
}

- (void)displayTargetTotalTime
{
    NSString * timeString;
    
    if(mLap >= TOTAL_LAP) {
        return;
    }
    if(mLap >= REF_LAP) {
        // Gesamtzeit Wertungsprüfung auf Basis der Referenzrunde
        NSTimeInterval timeInterval = (TOTAL_LAP-REF_LAP) * refLapTimeInterval;
        // Zeit seit Ende der Referenzrunde abziehen
        NSDate *currentDate = [NSDate date];
        timeInterval -= [currentDate timeIntervalSinceDate:endRefTime];
        // Niedriger Ton 5 Sekunden vor Ende
        if(((timeInterval < 5.05) && (timeInterval > 4.95)) ||
           ((timeInterval < 4.05) && (timeInterval > 3.95)) ||
           ((timeInterval < 3.05) && (timeInterval > 2.95)) ||
           ((timeInterval < 2.05) && (timeInterval > 1.95)) ||
           ((timeInterval < 1.05) && (timeInterval > 0.95)))
        {
            AudioServicesPlaySystemSound(mBeepLow);
        }
        // Hoher Ton am Ende
        if((timeInterval < 0.05) && (timeInterval > -0.05))
        {
            AudioServicesPlaySystemSound(mBeepHigh);
        }
        // Wieder hochzählen
        NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
        if(timeInterval < 0) {
            timeInterval = -timeInterval;
//            targetTotalTime.backgroundColor = [UIColor clearColor];
            [dateFormatter setDateFormat:@"-m:ss.S"];
        } else {
            [dateFormatter setDateFormat:@"m:ss.S"];
        }
        NSDate *timerDate = [NSDate dateWithTimeIntervalSince1970:timeInterval];
        [dateFormatter setTimeZone:[NSTimeZone timeZoneForSecondsFromGMT:0.0]];
        timeString = [dateFormatter stringFromDate:timerDate];
    } else {
        timeString = @"";
//        targetTotalTime.backgroundColor = [UIColor whiteColor];
    }
    targetTotalTime.text = timeString;
}

-(IBAction)inc:(id)sender
{
    mLap++;
    [self displayCounter];
    [self displayTargetLapTime];
    AudioServicesPlaySystemSound(mClick);
}

-(IBAction)lap:(id)sender
{
    if(!running) {
        running = true;
        startTime = [NSDate date];
        timer = [NSTimer scheduledTimerWithTimeInterval:1.0/10.0 target:self selector:@selector(displayTimerTimes) userInfo:nil repeats:YES];
    }
    mLap++;
    if(mLap == REF_LAP-1) {
        // Referenzrunde
        startRefTime = [NSDate date];
    }
    if(mLap == REF_LAP) {
        endRefTime = [NSDate date];
        [self calculateRefTime];
        [self displayRefTime];
        incButton.enabled = true;
        decButton.enabled = true;
    }
    [self displayCounter];
    [self displayLastLapTime];
    [self displayTargetLapTime];
    startLapTime = [NSDate date];
    AudioServicesPlaySystemSound(mClick);
}

-(IBAction)reset:(id)sender
{
    UIAlertView *alert = [[UIAlertView alloc]
                          initWithTitle:@"Reset" message:@"Rundenzähler und Timer zurücksetzen?" delegate:self cancelButtonTitle:@"Nein" otherButtonTitles:@"Ja", nil];
    alert.tag = ALERT_RESET;
    [alert show];
}

-(void)reset
{
    running = false;
    inBox = false;
    boxButton.enabled = true;
    [self resetCounter];
    [self resetTimer];
    incButton.enabled = false;
    decButton.enabled = false;
}

-(void)resetCounter
{
    mLap = -1;
    [self displayCounter];
}

-(void)resetTimer
{
    if(timer != nil) {
        [timer invalidate];
        timer = nil;
    }
    startTime = nil;
    startRefTime = nil;
    endRefTime = nil;
    startLapTime = nil;
    refLapTimeInterval = 0;
    [self displayTotalTime];
    [self displayLapTime];
    [self displayRefTime];
    [self displayLastLapTime];
    [self displayTargetLapTime];
    [self displayTargetTotalTime];
}

- (void)viewDidLoad
{
    [super viewDidLoad];
	// Do any additional setup after loading the view, typically from a nib.
    // Create the sound ID
    NSURL* clickURL = [[NSBundle mainBundle]
                       URLForResource:@"Default" withExtension:@"wav"];
    AudioServicesCreateSystemSoundID((__bridge CFURLRef)clickURL, &mClick);
    NSURL* beepLowURL = [[NSBundle mainBundle]
                         URLForResource:@"Beep-Low" withExtension:@"wav"];
    AudioServicesCreateSystemSoundID((__bridge CFURLRef)beepLowURL, &mBeepLow);
    NSURL* beepHighURL = [[NSBundle mainBundle]
                          URLForResource:@"Beep-High" withExtension:@"wav"];
    AudioServicesCreateSystemSoundID((__bridge CFURLRef)beepHighURL, &mBeepHigh);
    [self resetTimer];
    [[UIApplication sharedApplication] setIdleTimerDisabled:YES];
}

- (void)viewDidUnload
{
    [super viewDidUnload];
	// Do any additional setup after loading the view, typically from a nib.
    AudioServicesDisposeSystemSoundID(mClick);
    AudioServicesDisposeSystemSoundID(mBeepLow);
    AudioServicesDisposeSystemSoundID(mBeepHigh);
    [[UIApplication sharedApplication] setIdleTimerDisabled:NO];
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)viewDidLayoutSubviews {
    [super viewDidLayoutSubviews];

    // Design-Größe (iPad-Portrait)
    CGFloat designWidth = 768.0;
    CGFloat designHeight = 1024.0;

    // Tatsächliche Bildschirmgröße
    CGSize screenSize = self.view.bounds.size; // Das ist *aktuelle Ausrichtung*
    CGFloat screenWidth = screenSize.width;
    CGFloat screenHeight = screenSize.height;

    if (round(screenWidth) != round(designWidth)) {
        // Einheitliche Skalierung berechnen
        CGFloat scaleX = screenWidth / designWidth;
        CGFloat scaleY = screenHeight / designHeight;
        CGFloat scale = MIN(scaleX, scaleY); // damit nichts rausfällt

        // Skalierung anwenden
        self.view.transform = CGAffineTransformMakeScale(scale, scale);

        // Nach Skalierung: neue Größe
        CGFloat scaledWidth = designWidth * scale;
        CGFloat scaledHeight = designHeight * scale;

        // Differenz berechnen, um zu zentrieren
        CGFloat dx = (screenWidth - scaledWidth) / 2.0;
        CGFloat dy = (screenHeight - scaledHeight) / 2.0;

        // Frame verschieben zur Zentrierung
        self.view.frame = CGRectMake(dx, dy, scaledWidth, scaledHeight);
    }

}

@end

File: ChallengeSwift/ViewController.swift
//
//  ViewController.swift
//  ChallengeSwift
//
//  Created by Vibe Code on 2024.
//

import UIKit
import AVFoundation

class ViewController: UIViewController {

    // MARK: - IBOutlets
    @IBOutlet weak var counter: UILabel!
    @IBOutlet weak var totalTime: UITextField!
    @IBOutlet weak var lapTime: UITextField!
    @IBOutlet weak var lastLapTime: UITextField!
    @IBOutlet weak var refLapTime: UITextField!
    @IBOutlet weak var targetLapTime: UITextField!
    @IBOutlet weak var targetTotalTime: UITextField!
    @IBOutlet weak var boxButton: UIButton!
    @IBOutlet weak var incButton: UIButton!
    @IBOutlet weak var decButton: UIButton!
    @IBOutlet weak var refPlusHButton: UIButton!
    @IBOutlet weak var refMinusHButton: UIButton!
    @IBOutlet weak var refPlusZButton: UIButton!
    @IBOutlet weak var refMinusZButton: UIButton!

    // MARK: - Properties
    private var mLap: Int = -1                      // Rundenanzahl
    private var running: Bool = false               // Gibt an, ob die Stopuhr läuft
    private var inBox: Bool = false                 // Gibt an, ob der Boxenstop bereits erfolgt ist
    private var timer: Timer?                     // Timer für die asynchrone Zeitanzeige
    
    private var startTime: Date?                  // Startzeit der Messung
    private var startLapTime: Date?               // Startzeit der aktuellen Runde
    private var startRefTime: Date?               // Startzeit der Referenzrunde
    private var endRefTime: Date?                 // Endzeit der Referenzrunde
    
    // Referenzrunde
    private var refLapTimeInterval: TimeInterval = 0  // Zeit der Referenzrunde

    // MARK: - View Lifecycle
    override func viewDidLoad() {
        super.viewDidLoad()
        
        // Sound-Manager initialisieren
        _ = SoundManager.shared
        
        // Timer zurücksetzen
        resetTimer()
        
        // Bildschirm-Sperre deaktivieren
        UIApplication.shared.isIdleTimerDisabled = true
        
        // Button-States initialisieren
        incButton.isEnabled = false
        decButton.isEnabled = false
    }

    override func viewDidDisappear(_ animated: Bool) {
        super.viewDidDisappear(animated)
        
        // Sound-Manager bereinigen
        SoundManager.shared.dispose()
        UIApplication.shared.isIdleTimerDisabled = false
    }

    override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()

        // Design-Größe (iPad-Portrait)
        let designWidth: CGFloat = 768.0
        let designHeight: CGFloat = 1024.0

        // Tatsächliche Bildschirmgröße
        let screenSize = self.view.bounds.size
        let screenWidth = screenSize.width
        let screenHeight = screenSize.height

        if round(screenWidth) != round(designWidth) {
            // Einheitliche Skalierung berechnen
            let scaleX = screenWidth / designWidth
            let scaleY = screenHeight / designHeight
            let scale = min(scaleX, scaleY) // damit nichts rausfällt

            // Skalierung anwenden
            self.view.transform = CGAffineTransform(scaleX: scale, y: scale)

            // Nach Skalierung: neue Größe
            let scaledWidth = designWidth * scale
            let scaledHeight = designHeight * scale

            // Differenz berechnen, um zu zentrieren
            let dx = (screenWidth - scaledWidth) / 2.0
            let dy = (screenHeight - scaledHeight) / 2.0

            // Frame verschieben zur Zentrierung
            self.view.frame = CGRect(x: dx, y: dy, width: scaledWidth, height: scaledHeight)
        }
    }

    // MARK: - IBActions
    @IBAction func lap(_ sender: Any) {
        if !running {
            running = true
            startTime = Date()
            timer = Timer.scheduledTimer(withTimeInterval: 0.1, repeats: true) { [weak self] _ in
                self?.displayTimerTimes()
            }
        }
        mLap += 1
        if mLap == REF_LAP - 1 {
            // Referenzrunde
            startRefTime = Date()
        }
        if mLap == REF_LAP {
            endRefTime = Date()
            calculateRefTime()
            displayRefTime()
            incButton.isEnabled = true
            decButton.isEnabled = true
        }
        displayCounter()
        displayLastLapTime()
        displayTargetLapTime()
        startLapTime = Date()
        SoundManager.shared.playClick()
    }

    @IBAction func box(_ sender: Any) {
        SoundManager.shared.playClick()
        let alert = UIAlertController(title: "Box", message: "Boxenstop?", preferredStyle: .alert)
        alert.addAction(UIAlertAction(title: "Nein", style: .cancel, handler: nil))
        alert.addAction(UIAlertAction(title: "Ja", style: .default, handler: { [weak self] _ in
            self?.inBox = true
            self?.boxButton.isEnabled = false
        }))
        present(alert, animated: true, completion: nil)
    }

    @IBAction func reset(_ sender: Any) {
        let alert = UIAlertController(title: "Reset", message: "Rundenzähler und Timer zurücksetzen?", preferredStyle: .alert)
        alert.addAction(UIAlertAction(title: "Nein", style: .cancel, handler: nil))
        alert.addAction(UIAlertAction(title: "Ja", style: .default, handler: { [weak self] _ in
            self?.reset()
        }))
        present(alert, animated: true, completion: nil)
    }

    @IBAction func inc(_ sender: Any) {
        mLap += 1
        displayCounter()
        displayTargetLapTime()
        SoundManager.shared.playClick()
    }

    @IBAction func dec(_ sender: Any) {
        mLap -= 1
        displayCounter()
        displayTargetLapTime()
        SoundManager.shared.playClick()
    }

    @IBAction func refPlusH(_ sender: Any) {
        refLapTimeInterval += 0.01
        displayNewRefTime()
    }

    @IBAction func refMinusH(_ sender: Any) {
        refLapTimeInterval -= 0.01
        displayNewRefTime()
    }

    @IBAction func refPlusZ(_ sender: Any) {
        refLapTimeInterval += 0.1
        displayNewRefTime()
    }

    @IBAction func refMinusZ(_ sender: Any) {
        refLapTimeInterval -= 0.1
        displayNewRefTime()
    }

    // MARK: - Display Methods
    private func displayNewRefTime() {
        SoundManager.shared.playClick()
        displayRefTime()
        displayCounter()
        displayTargetLapTime()
        displayTargetTotalTime()
    }

    private func displayCounter() {
        let lap = mLap == -1 ? 0 : mLap
        counter.text = "\(lap)"
    }

    private func displayLapTime() {
        var timeString = ""
        if startTime != nil, let currentDate = startLapTime {
            let timeInterval = Date().timeIntervalSince(currentDate)
            let timerDate = Date(timeIntervalSince1970: timeInterval)
            let dateFormatter = DateFormatter()
            dateFormatter.dateFormat = "m:ss.S"
            dateFormatter.timeZone = TimeZone(secondsFromGMT: 0)
            timeString = dateFormatter.string(from: timerDate)
        }
        lapTime.text = timeString
    }

    private func displayLastLapTime() {
        var timeString = ""
        if let currentDate = startLapTime {
            let timeInterval = Date().timeIntervalSince(currentDate)
            let timerDate = Date(timeIntervalSince1970: timeInterval)
            let dateFormatter = DateFormatter()
            dateFormatter.dateFormat = "m:ss.S"
            dateFormatter.timeZone = TimeZone(secondsFromGMT: 0)
            timeString = dateFormatter.string(from: timerDate)
        }
        lastLapTime.text = timeString
    }

    private func calculateRefTime() {
        if let start = startRefTime, let end = endRefTime {
            refLapTimeInterval = end.timeIntervalSince(start)
        }
    }

    private func displayRefTime() {
        var timeString = ""
        if refLapTimeInterval != 0 {
            let timerDate = Date(timeIntervalSince1970: refLapTimeInterval)
            let dateFormatter = DateFormatter()
            dateFormatter.dateFormat = "m:ss.SS"
            dateFormatter.timeZone = TimeZone(secondsFromGMT: 0)
            timeString = dateFormatter.string(from: timerDate)
        }
        refLapTime.text = timeString
    }

    private func displayTimerTimes() {
        displayTotalTime()
        displayTargetTotalTime()
        displayLapTime()
    }

    private func displayTotalTime() {
        var timeString = "0:00.0"
        if let start = startTime {
            let timeInterval = Date().timeIntervalSince(start)
            let timerDate = Date(timeIntervalSince1970: timeInterval)
            let dateFormatter = DateFormatter()
            dateFormatter.dateFormat = "m:ss.S"
            dateFormatter.timeZone = TimeZone(secondsFromGMT: 0)
            timeString = dateFormatter.string(from: timerDate)
        }
        totalTime.text = timeString
    }

    private func displayTargetLapTime() {
        var timeString = ""
        
        if mLap >= REF_LAP {
            // Gesamtzeit Wertungsprüfung auf Basis der Referenzrunde
            var timeInterval = TimeInterval((TOTAL_LAP - REF_LAP) * Int(refLapTimeInterval))
            // Zeit seit Ende der Referenzrunde abziehen
            if let endRef = endRefTime {
                timeInterval -= Date().timeIntervalSince(endRef)
            }
            // Vor Boxenstopp? --> Boxenzeit abziehen
            if !inBox {
                timeInterval -= TimeInterval(BOX_TIME)
            }
            // Durch die Anzahl verbleibender Runden teilen
            timeInterval = timeInterval / TimeInterval((TOTAL_LAP - mLap))
            let timerDate = Date(timeIntervalSince1970: timeInterval)
            let dateFormatter = DateFormatter()
            dateFormatter.dateFormat = "m:ss.S"
            dateFormatter.timeZone = TimeZone(secondsFromGMT: 0)
            timeString = dateFormatter.string(from: timerDate)
        }
        targetLapTime.text = timeString
    }

    private func displayTargetTotalTime() {
        var timeString = ""
        
        if mLap >= TOTAL_LAP {
            return
        }
        if mLap >= REF_LAP {
            // Gesamtzeit Wertungsprüfung auf Basis der Referenzrunde
            var timeInterval = TimeInterval((TOTAL_LAP - REF_LAP) * Int(refLapTimeInterval))
            // Zeit seit Ende der Referenzrunde abziehen
            if let endRef = endRefTime {
                timeInterval -= Date().timeIntervalSince(endRef)
            }
            
            // Niedriger Ton 5 Sekunden vor Ende
            let beepTimes = [5.0, 4.0, 3.0, 2.0, 1.0]
            for beepTime in beepTimes {
                if timeInterval < beepTime + 0.05 && timeInterval > beepTime - 0.05 {
                    SoundManager.shared.playBeepLow()
                    break
                }
            }
            
            // Hoher Ton am Ende
            if timeInterval < 0.05 && timeInterval > -0.05 {
                SoundManager.shared.playBeepHigh()
            }
            
            // Wieder hochzählen
            let dateFormatter = DateFormatter()
            if timeInterval < 0 {
                timeInterval = -timeInterval
                dateFormatter.dateFormat = "-m:ss.S"
            } else {
                dateFormatter.dateFormat = "m:ss.S"
            }
            let timerDate = Date(timeIntervalSince1970: timeInterval)
            dateFormatter.timeZone = TimeZone(secondsFromGMT: 0)
            timeString = dateFormatter.string(from: timerDate)
        }
        targetTotalTime.text = timeString
    }

    // MARK: - Reset Methods
    private func reset() {
        running = false
        inBox = false
        boxButton.isEnabled = true
        resetCounter()
        resetTimer()
        incButton.isEnabled = false
        decButton.isEnabled = false
    }

    private func resetCounter() {
        mLap = -1
        displayCounter()
    }

    private func resetTimer() {
        timer?.invalidate()
        timer = nil
        startTime = nil
        startRefTime = nil
        endRefTime = nil
        startLapTime = nil
        refLapTimeInterval = 0
        displayTotalTime()
        displayLapTime()
        displayRefTime()
        displayLastLapTime()
        displayTargetLapTime()
        displayTargetTotalTime()
    }
}
